articles/115_topic_and_service_name_mapping.md

```diff
@@ -0,0 +1,423 @@
+author: '[William Woodall](https://github.com/wjwwood)'
+published: true
+---
+
+- This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Context
+
+Before proposing constraints for ROS 2 topic and service names and a mapping to underlying DDS topics, this article will first summarize the existing guidelines and restrictions for both ROS 1 and DDS.
+
+### ROS 1 Names
+
+In ROS 1, topic and service name guidelines are all covered under the umbrella of "ROS Names" and have these restrictions:
+
+*From [http://wiki.ros.org/Names](http://wiki.ros.org/Names):*
+
+```
+A valid name has the following characteristics:
+
+* First character is an alpha character ([a-z|A-Z]), tilde (~) or forward slash (/)
+* Subsequent characters can be alphanumeric ([0-9|a-z|A-Z]), underscores (_), or forward slashes (/)
+
+Exception: base names (described below) cannot have forward slashes (/) or tildes (~) in them.
+```
+
+### DDS Topic Names
+
+In DDS, topic names are restricted by these restrictions:
+
+*From DDS 1.4 specification, or the [RTI documentation](http://community.rti.com/rti-doc/510/ndds/doc/html/api_cpp/group__DDSQueryAndFilterSyntaxModule.html):*
+
+```
+TOPICNAME - A topic name is an identifier for a topic, and is defined as any series of characters 'a', ..., 'z', 'A', ..., 'Z', '0', ..., '9', '_' but may not start with a digit.
+```
+
+*Note:* that the DDS specification has a known typo, where it says `-` are allowed, but the RTI documentation correctly lists `_` as allowed.
+
+Additionally, DDS has a hard limit on topic names of 256 characters, so an additional goal is to minimize the number of extra characters used when mapping from ROS to DDS names.
+
+## ROS 2 Topic and Service Name Constraints
+
+In this section an outline of the proposed constrains for ROS 2 topic and service names will be enumerated along with rationales where appropriate.
+
+For convenience here is a summary of all rules for topic and service names in ROS 2:
+
+- may contain alphanumeric characters (`[0-9|a-z|A-Z]`), underscores (`_`), or forward slashes (`/`)
+- may use balanced curly braces (`{}`) for substitutions
+- may start with a tilde (`~`), the private namespace substitution character
+- must not start with a numeric character (`[0-9]`)
+- must not contain any number of repeated underscores (`_`)
+- must not end with an underscore (`_`)
+- must not have an underscore (`_`) followed by a forward slash (`/`), i.e. `_/`
+- must not end with a forward slash (`/`)
+- must not contain any number of repeated forward slashes (`/`)
+- must separate a tilde (`~`) from the rest of the name with a forward slash (`/`), i.e. `~/foo` not `~foo`
+- must have balanced curly braces (`{}`) when used, i.e. `{sub}/foo` but not `{sub/foo` nor `/foo}`
```

Comment:
Yeah, I thought about that, but since you have to do the substitutions before checking for correctness (as mentioned in the document) I didn't think it would be a burden to be slightly less restrictive for the substitution name.

I think you have to do the substitution before validating the topic name because if the substituted value has a leading underscore, or has a trailing underscore and is followed by an underscore, then the expanded topic name could become invalid where the original one was not. I think you also cannot just replace substitutions with empty string either because an example like this `/{name}_foo` would not be valid if the substitution is just ignored to do pre-expansion validation.

The only case where not permitting `__` in the substitution text would be useful is a pre-expansion check which naively asserts that no double underscore (`__`) is contained in the name string. I'm just not sure how useful that check is, since you need to do it again after expanding.

Having said all that, I'm fine with not allowing `__` in the substitution name because I don't think it is something we need to support.